src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -229,39 +231,60 @@ internal ItemManager(RecentItemsManager recentItemsManager, IGlobalOptionService
-                    : ImmutableArray<CompletionFilterWithState>.Empty;
+                if (!filteringResult.HasValue)
+                    return null;
 
-                // If this was deletion, then we control the entire behavior of deletion ourselves.
-                if (initialRoslynTriggerKind == CompletionTriggerKind.Deletion)
-                {
-                    return HandleDeletionTrigger(reason, initialListOfItemsToBeIncluded, filterText, updatedFilters, hasSuggestedItemOptions, highlightMatchingPortions, completionHelper);
-                }
+                var (selectedItemIndex, selectionHint, uniqueItem) = filteringResult.Value;
 
-                Func<ImmutableArray<(RoslynCompletionItem, PatternMatch?)>, string, ImmutableArray<RoslynCompletionItem>> filterMethod;
-                if (completionService == null)
+                // Editor is providing us a list of "default" items to consider for selection.
+                if (!data.Defaults.IsDefaultOrEmpty)
                 {
-                    filterMethod = (itemsWithPatternMatches, text) => CompletionService.FilterItems(completionHelper, itemsWithPatternMatches, text);
-                }
-                else
-                {
-                    Contract.ThrowIfNull(document);
-                    filterMethod = (itemsWithPatternMatches, text) => completionService.FilterItems(document, itemsWithPatternMatches, text);
+                    var tick = Environment.TickCount;
+
+                    var selectedItem = initialListOfItemsToBeIncluded[selectedItemIndex].RoslynCompletionItem;
+
+                    // "Preselect" is only used when we have high confidence with the selection, so don't override it.
+                    if (selectedItem.Rules.MatchPriority < MatchPriority.Preselect)
+                    {
+                        int defaultsMatchingIndex;
+                        var useAggressiveDefaultsMatching = session.TextView.Options.GetOptionValue<bool>(AggressiveDefaultsMatchingOptionName);
+
+                        if (useAggressiveDefaultsMatching)
+                        {
+                            defaultsMatchingIndex = GetAggressiveDefaultsMatch(initialListOfItemsToBeIncluded, data.Defaults);
+                            if (!hasSuggestedItemOptions && defaultsMatchingIndex >= 0)
+                                selectionHint = UpdateSelectionHint.Selected;
+                        }
+                        else
+                        {
+                            defaultsMatchingIndex = GetDefaultsMatch(filterText, initialListOfItemsToBeIncluded, selectedItemIndex, data.Defaults);
+                        }
+
+                        if (defaultsMatchingIndex >= 0)
+                        {
+                            selectedItemIndex = defaultsMatchingIndex;
+                        }
+                    }
+
+                    AsyncCompletionLogger.LogGetDefaultsMatchTicksDataPoint(Environment.TickCount - tick);
```

Comment:
can you extract all this into a helpe rmethod.  this method is already insanely large.  (that extent, we might want a helper type here).